### PR TITLE
ByteToMessageDecoder must call fireChannelReadComplete if any channel…

### PR DESCRIPTION
**Pull request by normanmaurer**
…Read(...) produced data.

Motivation:

Commit d63bb4811ed8ccd5d9e45853f3ac6aee9da7ecab introduced a regression which could lead to missing calls of fireChannelReadComplete(). The problem could happen when we had multiple channelRead(...) calls and the last call of it not produced any data while the previous ones did. In this case we missed to call fireChannelReadComplete() as decodeWasNull was previous set to true while it should be false.

Modifications:

- Correctly only update decodeWasNull to true if no previous channelRead(...) call produced data.
- Add unit test.

Result:

Correctly call fireChannelReadComplete() in all cases.

### CODE REVIEW

File: codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java [hunk 1]
```diff
@@ -280,7 +280,11 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
                 }
 
                 int size = out.size();
-                decodeWasNull = !out.insertSinceRecycled();
+                if (decodeWasNull) {
+                    // Only update decodeWasNull if we did not produce data in a previous channelRead(...) call.
+                    decodeWasNull = !out.insertSinceRecycled();
```
Comment (Scottmitch) [inline codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java]: consider avoiding a conditional/branch statement here and just do the assignment.

```java
decodeWasNull = decodeWasNull && !out.insertSinceRecycled();
```

File: codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java [hunk 2]
```diff
@@ -344,4 +344,50 @@ public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
...
+                assertTrue(readCompleteExpected.get());
+                readCompleteCount.incrementAndGet();
+            }
+        });
+
+        channel.pipeline().fireChannelRead(Unpooled.wrappedBuffer(new byte[] {'a'}));
+        channel.pipeline().fireChannelReadComplete();
+        assertEquals(0, readCompleteCount.get());
+
+        channel.pipeline().fireChannelRead(Unpooled.wrappedBuffer(new byte[] {'b'}));
```
Comment (Scottmitch) [inline codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java]: add static import for `Unpooled.wrappedBuffer`?

File: codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java (hunk 1 above)
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java]: used bitwise at the end :)

**Comment by normanmaurer** [comment]
@Scottmitch actually after more debugging I think what we tried to archive with https://github.com/netty/netty/commit/d63bb4811ed8ccd5d9e45853f3ac6aee9da7ecab is not safe to do... I would prefer to revert https://github.com/netty/netty/commit/d63bb4811ed8ccd5d9e45853f3ac6aee9da7ecab for now and think more about it in general. WDYT ?

**Comment by normanmaurer** [comment]
Let me close this one for now... I will investigate

**Pull request closed without merging**